### DISCUSSION TITLE
Welcome to wifibroadcast Discussions!

### DISCUSSION THREAD

**Discussion by svpcom** [svpcom/wfb-ng#80] [General]
## Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by iamfaith** [comment]
Dear svpcom,

I was able to receive data with your repo between the internal wifi card of raspberry pi zero and rt28xx wifi cards.

![image](https://user-images.githubusercontent.com/16201784/107144991-a0859380-6979-11eb-9b04-122d584a77a9.png)

I just filter this address in Wireshark and catch the packet.

I have a question about a feature. I don't want encryption of video and I only use it for video transmitting. Can I remove the session key? I will this can reduce latency a lot.

**Reply by svpcom** [reply]
This will require a lot of changes in the code.

**Reply by iamfaith** [reply]
Yes, I understand. But what if we keep the session key but don't encrypt data like this:
![image](https://user-images.githubusercontent.com/16201784/107164313-81264f00-69e9-11eb-80bd-2cdaf1ce92aa.png)
Can I send it like this? I am not sure what wblock_hdr_t is.

**Reply by svpcom** [reply]
Why you need to disable encryption? Default camera to display latency is ~ 100ms and the most of it is video encoder latency + video decoder + display sync (up to 33ms for 30fps). WFB latency < 5ms (telemetry/latency_test.py). Encryption is packet-based so it doesn't introduce any measurable latency.

**Reply by iamfaith** [reply]
> Why you need to disable encryption? Default camera to display latency is ~ 100ms and the most of it is video encoder latency + video decoder + display sync (up to 33ms for 30fps). WFB latency < 5ms (telemetry/latency_test.py). Encryption is packet-based so it doesn't introduce any measurable latency.

Thanks for your reply. Actually, I would like to install it into an FPV drone. So I would try my best to reduce the latency. Is there any way to reduce latency from the airside? Replacing camera? I think 100ms for the default camera to display is too high. The raspivid command should have hardware acceleration.

Compared with 50ms latency in 480P@50FPS in the DIJ FPV system, I would like to achieve 80ms in the same resolution.

**Reply by svpcom** [reply]
In 30fps one frame is 33ms. Camera (in best case) will issue one frame in a time. Add encoder latency, decoder latency and display sync latency (1 frame)

**Reply by svpcom** [reply]
If you need to reduce latency - use 60fps or more

**Reply by iamfaith** [reply]
> If you need to reduce latency - use 60fps or more

Thanks for guiding me.

I am thinking will using more fps add more latency? I am afraid that too many packets for encoding and decoding will increase latency?

That's why I originally think that 30FPS or 40FPS is better.

**Reply by iamfaith** [reply]
Dear svpcom,

Does wifbroadcast support RX diversity? We may have 2 wifi adapters on the ground side. While we receive a data packet from two wifi cards,  how could it select one of them with the best RSSI respect to the TX?

**Reply by svpcom** [reply]
Yes. It supports RX and TX diversity out of box. In case of RX it accept packets from all cards and remove duplicates. In case of TX it use card with best RSSI of RX packets

**Reply by iamfaith** [reply]
--> In case of TX it use card with best RSSI of RX packets

How to achieve this? For example, I have two RX cards, and do I need to start two processes with the command of wfb_rx with the same radio port?

**Reply by iamfaith** [reply]
> Yes. It supports RX and TX diversity out of box. In case of RX it accept packets from all cards and remove duplicates. In case of TX it use card with best RSSI of RX packets

I think you are misunderstanding me. Could it support such scenarios such as we have two RX cards to receive packets from the same TX. But in some cases, the RX1 may lose some packets, but RX2 not. So we can combine the received packets together on the RX side(ground side). But it's different from what the single RX are receiving from multiple TX cards.

**Reply by svpcom** [reply]
You need only one wfb_rx process: wfb_rx accepts multiple wifi cards:
``` .bash
$ ./wfb_rx 
Local receiver: ./wfb_rx [-K rx_key] [-k RS_K] [-n RS_N] [-c client_addr] [-u client_port] [-p radio_port] [-l log_interval] interface1 [interface2] ...
Remote (forwarder): ./wfb_rx -f [-c client_addr] [-u client_port] [-p radio_port] interface1 [interface2] ...
Remote (aggregator): ./wfb_rx -a server_port [-K rx_key] [-k RS_K] [-n RS_N] [-c client_addr] [-u client_port] [-l log_interval]
Default: K='rx.key', k=8, n=12, connect=127.0.0.1:5600, radio_port=1, log_interval=1000
WFB version 21.2.4.28894-653b3c2c
```

https://github.com/svpcom/wifibroadcast/wiki/Setup-HOWTO

WFB support RX diversity via collection packets from multiple wifi cards. Also it support TX diversity (to select tx card based on rx rssi), but this requires python daemon (see HOWTO).  I'll recommend to use python daemon because it run out-of-box and configure all channels automatically. Using standalone wfb_rx and wfb_tx only recommended for developers who known WFB internals and known what they want to get.

**Comment by iamfaith** [comment]
Dear svpcom, 


I still can not figure out why the packet loss rate is so high. Same configuration for RX and  TX as discussed before. But without wifibroadcast using wifi and gstreamer, everything is ok

**Reply by svpcom** [reply]
1. Check that you use free channel (I'll recommend to use 5.8GHz band)
2. Check that you video bitrate < 1/2 of radio bitrate. For MCS#1 max video bitrate is 7mbit/s. I'll recommend to use 4mbit/s
Sample gstreamer pipeline for linux pc:
```
gst-launch-1.0 videotestsrc ! videoconvert ! video/x-raw,format=NV12,framerate=30/1,width=1280,height=720 ! timeoverlay ! vaapih264enc bitrate=4000 rate-control=cbr max-bframes=0 ! rtph264pay config-interval=1 ! udpsink host=127.0.0.1 port=5602
```
